### PR TITLE
Refactor read config

### PR DESCRIPTION
**Pull request by peterbryntesson**
feat: reading configuration for opc events.

### CODE REVIEW

**Comment by peterbryntesson** [comment]
We need to handle some merge issues here. PR will be updated soon.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedEventItemsModelEx.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        /// <param name="dataItems"></param>
+        /// <returns></returns>
+        public static IEnumerable<MonitoredItemModel> ToMonitoredItems(
+            this PublishedEventItemsModel dataItems) {
+            if (dataItems?.PublishedData == null) {
+                return Enumerable.Empty<MonitoredItemModel>();
+            }
+
+            var map = new Dictionary<string, MonitoredItemModel>();
+            foreach (var item in dataItems.PublishedData) {
```
Comment (buzzfrog) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedEventItemsModelEx.cs:29]: should the parameter be named dataItems even as we work with events?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs [hunk 2]
```diff
@@ -232,8 +263,36 @@ public class PublishedNodesJobConverter {
...
                 }
             }
+
+            if (item.OpcEvents != null) {
+                foreach (var node in item.OpcEvents) {
+                    if (string.IsNullOrEmpty(node.Id)) {
+                        node.Id = node.ExpandedNodeId;
+                    }
+                    if (scaleTestCount == 1) {
+                        yield return node;
```
Comment (buzzfrog) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs:273]: Do I understand it correctly that we don't need to assign OpcPublishingInterval as on line #245 @simonjaeger

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedEventItemsModelEx.cs (hunk 1 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedEventItemsModelEx.cs:29]: Good catch

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs (hunk 2 above)
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs:273]: Let me check @buzzfrog

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Data/OpcDataNodeModel.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        /// OpcSamplingInterval as TimeSpan.
+        /// </summary>
+        [IgnoreDataMember]
+        public TimeSpan? OpcSamplingIntervalTimespan {
+            get => OpcSamplingInterval.HasValue ?
+                TimeSpan.FromMilliseconds(OpcSamplingInterval.Value) : (TimeSpan?)null;
+            set => OpcSamplingInterval = value != null ?
+                (int)value.Value.TotalMilliseconds : (int?)null;
+        }
+
```
Comment (buzzfrog) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Data/OpcDataNodeModel.cs:26]: remove empty line

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterAttributeModel.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+﻿namespace Microsoft.Azure.IIoT.OpcUa.Publisher.Config.Models.Events {
+    using System.Runtime.Serialization;
+
+    /// <summary>
+    /// Class to describe the AttributeOperand.
```
Comment (buzzfrog) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterAttributeModel.cs:5]: change to FilterAttributeModel

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterSimpleAttributeModel.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿namespace Microsoft.Azure.IIoT.OpcUa.Publisher.Config.Models.Events {
+    using System.Collections.Generic;
+    using System.Runtime.Serialization;
+
+    /// <summary>
+    /// Class to describe the SimpleAttributeOperand.
```
Comment (buzzfrog) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterSimpleAttributeModel.cs:6]: let the comment reflect the class name

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs (hunk 2 above)
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesJobConverter.cs:274]: Fixed.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterSimpleAttributeModel.cs (hunk 5 above)
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterSimpleAttributeModel.cs:6]: Comment is actually accurate. The class FilterSimpleAttributeModel describes SimpleAttributeOperand in the spec. I'm more inclined to change the class name according to spec. We have more of these issues however that I have been holding off on. Thoughts?

Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterSimpleAttributeModel.cs:6]: Why can't we use SimpleAttributeOperandModel here instead? This is a duplicate.

Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/FilterSimpleAttributeModel.cs:6]: fix: reused existing model classes

**Comment by koepalex** [comment]
> namespace Microsoft.Azure.IIoT.OpcUa.Publisher.Config.Models.Data {

Copyright header is missing

---
Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Data/OpcDataNodeModel.cs:1 in bf5328b. [](commit_id = bf5328bd80077f46c44aaf07d45bf2ca937a8faf, deletion_comment = False)

**Comment by koepalex** [comment]
> namespace Microsoft.Azure.IIoT.OpcUa.Publisher.Config.Models.Events {

Copyright header is missing

---
Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/OpcEventNodeModel.cs:1 in bf5328b. [](commit_id = bf5328bd80077f46c44aaf07d45bf2ca937a8faf, deletion_comment = False)

**Comment by koepalex** [comment]
> namespace Microsoft.Azure.IIoT.OpcUa.Publisher.Config.Models {

Copyright header is missing

---
Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/OpcBaseNodeModel.cs:1 in bf5328b. [](commit_id = bf5328bd80077f46c44aaf07d45bf2ca937a8faf, deletion_comment = False)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Extensions/PublishedDataSetSourceModelEx.cs [hunk 6]
```diff
@@ -42,11 +42,13 @@ public static class PublishedDataSetSourceModelEx {
                 model.Connection?.User?.Type.ToString() +
                 model.Connection?.User?.Value.ToJson() +
                 model.SubscriptionSettings?.PublishingInterval.ToString() +
-                model.PublishedVariables.PublishedData.First()?.Id +
-                model.PublishedVariables.PublishedData.First()?.PublishedVariableNodeId +
-                model.PublishedVariables.PublishedData.First()?.PublishedVariableDisplayName +
-                model.PublishedVariables.PublishedData.First()?.SamplingInterval +
-                model.PublishedVariables.PublishedData.First()?.HeartbeatInterval;
+                model.PublishedVariables.PublishedData.FirstOrDefault()?.Id +
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Extensions/PublishedDataSetSourceModelEx.cs:45]: please optimize and store the first PublishedData into local variable

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs [hunk 7]
```diff
@@ -1253,6 +1253,258 @@ private class StandaloneIdentity : IIdentity {
...
+            Assert.Single(jobs.Single().WriterGroup.DataSetWriters.Single().DataSet.DataSetSource.PublishedEvents.PublishedData);
+            Assert.NotNull(jobs.Single().WriterGroup.DataSetWriters.Single().DataSet.DataSetSource.PublishedVariables);
+            Assert.NotNull(jobs.Single().WriterGroup.DataSetWriters.Single().DataSet.DataSetSource.PublishedEvents);
+            Assert.Empty(jobs.Single().WriterGroup.DataSetWriters.Single().DataSet.DataSetSource.PublishedVariables.PublishedData);
+            Assert.Single(jobs.Single().WriterGroup.DataSetWriters.Single().DataSet.DataSetSource.PublishedEvents.PublishedData);
+
+            var model = jobs.Single().WriterGroup.DataSetWriters.Single().DataSet.DataSetSource.PublishedEvents.PublishedData.Single();
+            Assert.Equal("i=2258", model.Id);
+
+            // TODO: ...
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs:1310]: please check, PR should have solved all TODO

Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs:1310]: maybe checking for Operator, Operands, BrowsePaths etc.

---
In reply to: [596731348](https://github.com/Azure/Industrial-IoT/pull/1063#discussion_r596731348) [](ancestors = 596731348)

Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs:1310]: Checking with @Dag, but I agree.

Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs:1310]: I'll take this and fix it.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Extensions/PublishedDataSetSourceModelEx.cs (hunk 6 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Extensions/PublishedDataSetSourceModelEx.cs]: Fixed it.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs (hunk 7 above)
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Storage/PublishedNodesJobConverterTests.cs]: Fixed.

**Review by koepalex — APPROVED** [review]
:shipit:

**Pull request merged**